### PR TITLE
New /from-stdin

### PR DESCRIPTION
**Pull request by akallabeth**
* Adds ```/from-stdin``` to ignore missing ```user```, ```domain``` and ```password```
* Implemented ```GatewayAuthenticate``` callbacks for windows and X11 clients
* Implemented full ```Authenticate``` callback for X11

### CODE REVIEW

File: client/common/cmdline.c [hunk 1]
```diff
@@ -169,6 +169,7 @@ COMMAND_LINE_ARGUMENT_A args[] =
 	{ "multitransport", COMMAND_LINE_VALUE_BOOL, NULL, BoolValueFalse, NULL, -1, NULL, "Support multitransport protocol" },
 	{ "assistance", COMMAND_LINE_VALUE_REQUIRED, "<password>", NULL, NULL, -1, NULL, "Remote assistance password" },
 	{ "encryption-methods", COMMAND_LINE_VALUE_REQUIRED, "<40,56,128,FIPS>", NULL, NULL, -1, NULL, "RDP standard security encryption methods" },
+	{ "from-stdin", COMMAND_LINE_VALUE_FLAG, NULL, NULL, NULL, -1, NULL, "print version" },
```
Comment (bmiklautz) [inline client/common/cmdline.c]: I guess the help should be different than "print version"

**Comment by mfleisz** [comment]
+1

File: client/X11/xf_client.c [hunk 2]
```diff
@@ -1228,18 +1228,74 @@ static void xf_post_disconnect(freerdp* instance)
...
+			*username = StrSep(username, "\r");
+			*username = StrSep(username, "\n");
+		}
+	}
+
+	if (!*domain)
+	{
+		size_t domain_size = 0;
+		printf("%s", prompt[1]);
+		getline(domain, &domain_size, stdin);
```
Comment (MartinHaimberger) [inline client/X11/xf_client.c]: result of getline should be checked and on failture the buffers should be freed (see doc)

 If *lineptr is set to NULL and *n is set 0 before the call, then
       getline() will allocate a buffer for storing the line.  This buffer
       should be freed by the user program even if getline() failed.

File: client/X11/xf_client.c [hunk 3]
```diff
@@ -1228,18 +1228,74 @@ static void xf_post_disconnect(freerdp* instance)
...
 
-	if (freerdp_passphrase_read("Password: ", *password, password_size, instance->settings->CredentialsFromStdin) == NULL)
+	if (!username || !password || !domain)
 		return FALSE;
 
+	if (!*username)
+	{
+		size_t username_size = 0;
+		printf("%s", prompt[0]);
+		getline(username, &username_size, stdin);
```
Comment (MartinHaimberger) [inline client/X11/xf_client.c]: result of getline should be checked and on failture the buffers should be freed (see doc)

If *lineptr is set to NULL and *n is set 0 before the call, then
getline() will allocate a buffer for storing the line. This buffer
should be freed by the user program even if getline() failed.

**Comment by MartinHaimberger** [comment]
besides my remarks LGTM

**Comment by bmiklautz** [comment]
Although I'm very happy we have that again this PR unfortunately changes the default behavior. Usually , when no additional parameter were given, only the password was prompted. Now the domain is prompted as well. But for example if the username is given the Domain isn't asked for.
I'm not sure if we wan't need this in that form.  IMHO we should have a consistent way.
```
xfreerdp /v:127.0.0.1
[11:32:10:651] [25677:25678] [INFO][com.freerdp.client.x11] - No user name set. - Using login name:  xxxx
Domain:    
Password: 
```

**Pull request merged**